s3_credentials/cli.py

```diff
@@ -511,6 +510,82 @@ def iterate():
     )
 
 
+@cli.command()
+@click.argument("role_names", nargs=-1)
+@click.option("--details", help="Include attached policies (slower)", is_flag=True)
+@common_output_options
+@common_boto3_options
+def list_roles(role_names, details, nl, csv, tsv, **boto_options):
+    "List all roles"
+    iam = make_client("iam", **boto_options)
+    paginator = iam.get_paginator("list_roles")
+
+    headers = (
+        "Path",
+        "RoleName",
+        "RoleId",
+        "Arn",
+        "CreateDate",
+        "AssumeRolePolicyDocument",
+        "Description",
+        "MaxSessionDuration",
+        "PermissionsBoundary",
+        "Tags",
+        "RoleLastUsed",
+    )
+
+    if details:
+        headers += ("inline_policies", "attached_policies")
+
+    def iterate():
+        for response in paginator.paginate():
+            for role in response["Roles"]:
+                if role_names and role["RoleName"] not in role_names:
+                    continue
+                if details:
+                    role_name = role["RoleName"]
+                    role["inline_policies"] = []
+                    # Get inline policy names, then policy for each one
+                    role_policies_response = iam.list_role_policies(
+                        RoleName=role_name,
+                    )
+                    # TODO: Warn on IsTruncated, maybe even paginate
```

Comment:
I need a general solution for pagination. Maybe something like this:
```python
for policy_name in paginate(iam, "list_role_policies", "PolicyNames", RoleName=role_name):
    # ...
```
The `paginate()` function would wrap this pattern, seen elsewhere:
```python
paginator = iam.get_paginator("list_user_policies")
for response in paginator.paginate(UserName=username):
    for policy_name in response["PolicyNames"]:
        # ...
```